crypto/fipsmodule/kbkdf/kbkdf.c

```diff
@@ -0,0 +1,178 @@
+// Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+// SPDX-License-Identifier: Apache-2.0 OR ISC
+
+// This implements the "KDF in Feedback Mode" described in section 4.2:
+// https://csrc.nist.gov/publications/detail/sp/800-108/rev-1/final
+
+#include <stdbool.h>
+
+#include <openssl/hmac.h>
+#include <openssl/kbkdf.h>
+
+#include "../../internal.h"
+#include "../service_indicator/internal.h"
+
+// Convert a 32-bit little-endian value into a 32-bit big-endian value.
+// AWS-LC only supports little-endian hosts.
+static uint32_t to_be32(uint32_t host)
+{
+    uint32_t value = 0;
+
+    value |= (host & 0xff000000) >> 24;
+    value |= (host & 0x00ff0000) >> 8;
+    value |= (host & 0x0000ff00) << 8;
+    value |= (host & 0x000000ff) << 24;
+
+    return value;
+}
+
+// Surprised min32/min64/etc. didn't end up in stdint.h...
+static inline size_t min(size_t a, size_t b) {
+    return a < b ? a : b;
+}
+
+static inline size_t max(size_t a, size_t b) {
+    return a > b ? a : b;
+}
+
+#ifndef BYTEBITS
+#define BYTEBITS 8
+#endif
+
+// Generates |out_len| bytes of keying material from |key_in|, |label|,
+// |context|, and |iv| using the given |digest|, and outputs the result to
+// |out_key|. It returns one on success and zero on error.
+//
+// The |use_counter| option enables the "KDF in Feedback Mode with Counter"
+// mode.
+OPENSSL_EXPORT int KBKDF_feedback(uint8_t *out_key, size_t out_len,
+                                  const EVP_MD *digest,
+                                  const uint8_t *key_in, size_t key_in_len,
+                                  const uint8_t *label, size_t label_len,
+                                  const uint8_t *context, size_t context_len,
+                                  const uint8_t *iv, size_t iv_len,
+                                  bool use_counter)
+{
+    // Sanity checking.
+    if (out_key == NULL || digest == NULL || key_in == NULL ||
+        out_len < 1 || key_in_len < 1 ||
+        (key_in == NULL && key_in_len > 0) ||
+        (label == NULL && label_len > 0) ||
+        (context == NULL && context_len > 0) ||
+        (iv == NULL && iv_len > 0)) {
+        return 0;
+    }
+
+    int retval = 0;
+    uint8_t *ki = NULL;
```

Comment:
More descriptive name maybe, or you following the spec?

Reply:
Following the spec, I'll add a comment about the terrible variable names.